OrbitVulkanLayer/SubmissionTrackerTest.cpp

```diff
@@ -0,0 +1,1411 @@
+                                             post_submit_time, pid, kTimestamp1, kTimestamp2);
+}
+
+TEST_F(SubmissionTrackerTest,
+       CanRetrieveCommandBufferTimestampsForACompleteSubmissionAtSecondPresent) {
+  EXPECT_TWO_NEXT_READY_QUERY_SLOT_CALLS();
+  EXPECT_CALL(dispatch_table, GetQueryPoolResults)
+      .WillOnce(Return(mock_get_query_pool_results_function_not_ready))
+      .WillRepeatedly(Return(mock_get_query_pool_results_function_all_ready));
+  std::vector<uint32_t> actual_reset_slots;
+  EXPECT_CALL(timer_query_pool, ResetQuerySlots).Times(1).WillOnce(SaveArg<1>(&actual_reset_slots));
+  orbit_grpc_protos::CaptureEvent actual_capture_event;
+  auto mock_enqueue_capture_event =
+      [&actual_capture_event](orbit_grpc_protos::CaptureEvent&& capture_event) {
+        actual_capture_event = std::move(capture_event);
+        return true;
+      };
+  EXPECT_CALL(*producer, EnqueueCaptureEvent).Times(1).WillOnce(Invoke(mock_enqueue_capture_event));
+
+  producer->StartCapture();
+  tracker.TrackCommandBuffers(device, command_pool, &command_buffer, 1);
+  tracker.MarkCommandBufferBegin(command_buffer);
+  tracker.MarkCommandBufferEnd(command_buffer);
+  pid_t pid = GetCurrentThreadId();
+  uint64_t pre_submit_time = MonotonicTimestampNs();
+  std::optional<internal::QueueSubmission> queue_submission_optional =
+      tracker.PersistCommandBuffersOnSubmit(1, &submit_info);
+  tracker.PersistDebugMarkersOnSubmit(queue, 1, &submit_info, queue_submission_optional);
+  uint64_t post_submit_time = MonotonicTimestampNs();
+  tracker.CompleteSubmits(device);
+  tracker.CompleteSubmits(device);
+
+  EXPECT_THAT(actual_reset_slots, UnorderedElementsAre(kSlotIndex1, kSlotIndex2));
+  EXPECT_SINGLE_COMMAND_BUFFER_SUBMISSION_EQ(actual_capture_event, pre_submit_time,
+                                             post_submit_time, pid, kTimestamp1, kTimestamp2);
+}
+
+TEST_F(SubmissionTrackerTest, StopCaptureBeforeSubmissionWillResetTheSlots) {
+  EXPECT_TWO_NEXT_READY_QUERY_SLOT_CALLS();
+  EXPECT_CALL(dispatch_table, GetQueryPoolResults).Times(0);
+  std::vector<uint32_t> actual_reset_slots;
+  EXPECT_CALL(timer_query_pool, ResetQuerySlots).Times(1).WillOnce(SaveArg<1>(&actual_reset_slots));
+
+  EXPECT_CALL(*producer, EnqueueCaptureEvent).Times(0);
+
+  producer->StartCapture();
+  tracker.TrackCommandBuffers(device, command_pool, &command_buffer, 1);
+  tracker.MarkCommandBufferBegin(command_buffer);
+  tracker.MarkCommandBufferEnd(command_buffer);
+  producer->StopCapture();
+  std::optional<internal::QueueSubmission> queue_submission_optional =
+      tracker.PersistCommandBuffersOnSubmit(1, &submit_info);
+  tracker.PersistDebugMarkersOnSubmit(queue, 1, &submit_info, queue_submission_optional);
+  tracker.CompleteSubmits(device);
+
+  EXPECT_THAT(actual_reset_slots, UnorderedElementsAre(kSlotIndex1, kSlotIndex2));
+}
+
+TEST_F(SubmissionTrackerTest, CanRetrieveCommandBufferTimestampsWhenNotCapturingAtPresent) {
```

Comment:
Remove "at present"?

Reply:
Hmm, I am a bit unsure. The point of this test is, that it captures during the command buffer setup and submission, but not at the point of the present.

Reply:
Ah, I get it now. Maybe "command buffer timestamps recorded when capturing can be retrieved when not capturing"? It's even longer but I think clearer.

Reply:
Done.